Add a shell script to auto-deploy a k8s cluster on Azure

sgibson91 opened the pull request:
## Summary

Addition of a shell script which will install CLIs for Azure, Kubernetes and Helm then automatically deploy a Kubernetes cluster on Azure using values from a `config.json` file created by the user.

## Updates in this PR

* Addition of `setup.sh`
* Addition of `read_config.py` to read in `config.json` created by the user 
* Update of README to include usage instructions on creating `config.json` and documenting what `setup.sh` will do

## Remaining TO DOs and Desirables

- [x] Check the installation scripts for the CLI's. Are these scripts OS version dependent? Eventually, I would like this script to become OS-neutral (e.g. set an "os" key in `config.json` to "mac", "linux" or "windows")
- [ ] On [line 53](https://github.com/sgibson91/binderhub-deploy/blob/c89254efb88b885504e9b9ebfe65a3906e7e4daa/setup.sh#L53) of `setup.sh`, I would like the script to wait until the STATUS of each k8s node is READY before continuing. This will depend on the value of `node_count` set by the user in `config.json`.

sgibson91 committed:
Add setup.sh script to install CLIs

Create setup.sh shell script to install the following command line
interfaces:
* Azure-CLI
* Kubernetes-CLI
* Helm-CLI

sgibson91 committed:
Update README to describe setup.sh shell script

* Describe the purpose of the project
* Begin a list of contained scripts
* Describe setup.sh script
* List docs for curl commands within setup.sh

sgibson91 committed:
Update setup.sh script to deploy a k8s cluster

* Add comments describing required inputs for the script
* Add commands to deploy k8s cluster onto Azure
* Add commands to configure helm and tiller on the cluster

sgibson91 committed:
Reformat setup script from shell to python, create config.json

* Reformat setup script from shell to python so that json file
  can be read in
* Create config.json template for Azure parameters
* Remove setup.sh
* Update README

sgibson91 committed:
Update README

* setup script is back to a shell script not python
* config.json file is read by read_config.py and parsed to setup.sh
* config.json has new format

sgibson91 committed:
Update format of config.json

Put all azure related key/values under azure level

sgibson91 committed:
Add read_config.py script to read config.json

Python script to read config.json and parse output to shell script

sgibson91 committed:
Add setup.sh to install CLIs and deploy k8s cluster

setup.sh is a shell script which installs CLIs for Azure,
Kubernetes and Helm and then deploys a Kubernetes cluster onto
Azure using values from config.json

sgibson91 committed:
Delete setup.py, replace with setup.sh

sgibson91 committed:
Remove config.json and update README

Remove config.json and add instructions to the README on how to
create it

sgibson91 committed:
Ignore config.json and generated SSH keys

sgibson91 committed:
Make config.json block markdown so the comments don't show up red

sgibson91 committed:
Note that ssh-keygen asks for a passphrase and this can be blank

sgibson91 committed:
Update .gitignore file to ignore .secret folder

Git ignore sensitive info stored in the .secret/ folder

sgibson91 committed:
Update setup shell script

Update setup.sh to not print output from az login command and
create a .secret folder for the SSH keys to be stored in

sgibson91 committed:
Add -p flag to mkdir in setup.sh: no error if folders exist

sgibson91 committed:
Update README

* Include instructions on making the shell scripts executable
* Include links to Azure resources for completing the config.json
  file

sgibson91 committed:
Make bash code-block markdown in README

sgibson91 committed:
Add link to nicain's binder-deploy repo

sgibson91 committed:
Merge branch 'dev/add-setup-script' of https://github.com/sgibson91/binderhub-deploy into dev/add-setup-script

jemrobinson reviewed (approved):
LGTM. Was there anything specific you wanted me to check?

sgibson91 commented:
Thanks. Let's meet next week and go through in a bit more detail 🙂

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
For portability, possibly use '#!/usr/bin/env bash' in case bash is in /usr/bin or somewhere else entirely random?

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+vm_size=${vararray[5]}
+
+# Install Azure-CLI
+curl -L https://aka.ms/InstallAzureCli
```
I think it would be worth considering wrapping this in a test to see if 'az' is already available on the system, with a successful check skipping the step, and a failure resulting in asking the user if they want to install via the OS package manager, or by a generic script. Perhaps with some rudimentary checks to see whether the user has an apt/yum/zypper-based packaging system and offering those options?

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+curl -L https://aka.ms/InstallAzureCli
+
+# Install kubectl - Kubernetes-CLI
+curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/darwin/amd64/kubectl
```
Similar comment to the az cli install - checking whether there's already kubectl available, and offering a package managed version if that's an option?

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+# Make the kubectl binary executable
+chmod +x ./kubectl
+# Move the binary into your PATH
+sudo mv ./kubectl /usr/local/bin/kubectl
```
This in particular looks like a possible can of worms if there hasn't been a check earlier on to see if there are other kubectl versions available :-)

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+# Install Helm cli - fetch install script and execute it locally
+curl https://raw.githubusercontent.com/helm/helm/master/scripts/get > get_helm.sh
+chmod 700 get_helm.sh
+./get_helm.sh
```
Again worth checking whether helm exists yet, and skipping this step if it does. To my surprise on a quick check I can't find a Ubuntu package for helm, though other environments (ie, OS X with brew) do have it at a system level.

tmbgreaves commented on setup.sh:
```diff
@@ -0,0 +1,69 @@
+az group create --name $res_grp_name --location $location --output table
+
+# Make a secret folder and a sub-folder for the cluster
+mkdir -p .secret && cd .secret && mkdir -p $cluster_name && cd $cluster_name
```
I think this condenses into a single 'mkdir -p .secret/${cluster_name}'

tmbgreaves committed:
Improving portability and packaging support

This commit:

* Improves portability of the setup script
* Uses OS packaging systems where possible to do installs
* Pulls in dependencies where they are not installed automatically
* Updates documentation

sgibson91 committed:
Merge pull request #2 from tmbgreaves/dev/add-setup-script

Make setup script more portable

sgibson91 committed:
Merge branch 'master' into dev/add-setup-script

sgibson91 committed:
Fix script links

sgibson91 committed:
Shorten az flags, Comment out ssh-keygen lines

ssh-keygen no longer required as now passing --generate-ssh-keys to az aks create command instead

sgibson91 commented:
@tmbgreaves if you want to check [`7644e7f`](https://github.com/alan-turing-institute/binderhub-deploy/pull/1/commits/7644e7f2157941630c082de50b47303e3657cfdf), I've commented out lines referencing `ssh-keygen` and swapped in `--generate-ssh-keys` command in `az aks create` line ([see these docs](https://docs.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-create)).

sgibson91 committed:
Correct generate ssh keys flag

tmbgreaves committed:
Removing all dependencies on/for ssh-keygen

tmbgreaves committed:
Change from python to jq for json parsing

tmbgreaves committed:
Make jq output raw data, not json-formatted data

tmbgreaves committed:
Suggested test for cluster nodes being ready

tmbgreaves committed:
Check tiller pod is running

sgibson91 committed:
Add set -e command to setup.sh

sgibson91 committed:
Merge pull request #3 from tmbgreaves/dev/add-setup-script

setup.sh script improvements
